# **Fake And Real News Classification**

---

Authors: [Dennis Kimiri](), [Femi Kamau](https://www.github.com/ctrl-Karugu), [Monicah Iwagit](), [Teofilo Gafna](), [Wendy Mwiti](https://www.github.com/WendyMwiti)

## 1. Business Understanding

## 2. Data Understanding

Load Libraries

In [29]:
# Data Manipulation
import pandas as pd
import numpy as np

# Data Visualisation
import matplotlib.pyplot as plt
import seaborn as sns

# Machine Learning
import sklearn
import nltk
import re

Load Data

In [30]:
# Load the real news data
real = pd.read_csv('./data/True.csv')

# Load the fake news data
fake = pd.read_csv('./data/Fake.csv')

In [31]:
# Preview real data
real.head()

title  \
0  As U.S. budget fight looms, Republicans flip t...   
1  U.S. military to accept transgender recruits o...   
2  Senior U.S. Republican senator: 'Let Mr. Muell...   
3  FBI Russia probe helped by Australian diplomat...   
4  Trump wants Postal Service to charge 'much mor...   

                                                text       subject  \
0  WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1  WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2  WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3  WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4  SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   

                 date  
0  December 31, 2017   
1  December 29, 2017   
2  December 31, 2017   
3  December 30, 2017   
4  December 29, 2017

In [32]:
# Preview the fake data
fake.head()

title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   

                date  
0  December 31, 2017  
1  December 31, 2017  
2  December 30, 2017  
3  December 29, 2017  
4  December 25, 2017

In [33]:
#  Add a column called 'category' to both DataFrames which will become our taget variable
real['category'] = 1
fake['category'] = 0

In [34]:
# Combine both DataFrames 
data = pd.concat([real, fake])

In [35]:
# Preview the new DataFrame
data

title  \
0      As U.S. budget fight looms, Republicans flip t...   
1      U.S. military to accept transgender recruits o...   
2      Senior U.S. Republican senator: 'Let Mr. Muell...   
3      FBI Russia probe helped by Australian diplomat...   
4      Trump wants Postal Service to charge 'much mor...   
...                                                  ...   
23476  McPain: John McCain Furious That Iran Treated ...   
23477  JUSTICE? Yahoo Settles E-mail Privacy Class-ac...   
23478  Sunnistan: US and Allied ‘Safe Zone’ Plan to T...   
23479  How to Blow $700 Million: Al Jazeera America F...   
23480  10 U.S. Navy Sailors Held by Iranian Military ...   

                                                    text       subject  \
0      WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1      WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2      WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3      WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4      SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   
...                                                  ...           ...   
23476  21st Century Wire says As 21WIRE reported earl...   Middle-east   
23477  21st Century Wire says It s a familiar theme. ...   Middle-east   
23478  Patrick Henningsen  21st Century WireRemember ...   Middle-east   
23479  21st Century Wire says Al Jazeera America will...   Middle-east   
23480  21st Century Wire says As 21WIRE predicted in ...   Middle-east   

                     date  category  
0      December 31, 2017          1  
1      December 29, 2017          1  
2      December 31, 2017          1  
3      December 30, 2017          1  
4      December 29, 2017          1  
...                   ...       ...  
23476    January 16, 2016         0  
23477    January 16, 2016         0  
23478    January 15, 2016         0  
23479    January 14, 2016         0  
23480    January 12, 2016         0  

[44898 rows x 5 columns]

In [36]:
class understanding(object):
    """A class that does basic Data Understanding"""
    
    def __init__(self, df):
        self.shape = df.shape
      # self.info = df.info()
        self.duplicates = df.duplicated().sum()
        self.missing = df.isna().sum()
        self.types = df.dtypes
    
    def counts(self, df, col):
        """ A function that gets the value counts of a column"""
        self.counts = df[col].value_counts()
        return self.counts
    
    def nunique(self, df, col):
        """ A function that gets the number of unique values of a column"""
        self.unique = df[col].nunique()
        return self.nunique()
        
    def unique(self, df, col):
        """ A function that gets the unique values of a column"""
        self.unique = df[col].unique()
        return self.unique

In [37]:
# Instantiate the class
data_understanding = understanding(data)

In [38]:
# Summary of the dataset
print(data_understanding.shape)
data.info()

(44898, 5)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 44898 entries, 0 to 23480
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   title     44898 non-null  object
 1   text      44898 non-null  object
 2   subject   44898 non-null  object
 3   date      44898 non-null  object
 4   category  44898 non-null  int64 
dtypes: int64(1), object(4)
memory usage: 2.1+ MB


From the summary above, we can see that the dataset contain 44898 rows and spans 5 columns. The columns are: title, text, subject, date and category. The category column is the target variable and the rest are the features.

Furthermore, the dataset contains 4 object columns and 1 integer column. The object columns are: title, text, subject and date. The integer column is: category (target variable). We may need to convert the date column to datetime format in the data preparation phase.

The dataset does not contain any missing values.

In [39]:
# Check for duplicates
print(f"Duplicates: {data_understanding.duplicates}")

Duplicates: 209


The dataset contains 209 duplicates. They shall be inspected and removed if necessary in the data preparation phase.

In [40]:
# Check the number of missing values
data_understanding.missing

title       0
text        0
subject     0
date        0
category    0
dtype: int64

In [41]:
# Inspect the subject column
data['subject'].value_counts()

politicsNews       11272
worldnews          10145
News                9050
politics            6841
left-news           4459
Government News     1570
US_News              783
Middle-east          778
Name: subject, dtype: int64

The articles that are within this dataset fall under 8 different subjects. These are: politicsNews, worldnews, News, politics, left-news, Government News, US_News, and Middle-east.

In [42]:
# Inspect the category column
data['category'].value_counts() / len(data) * 100

0    52.298543
1    47.701457
Name: category, dtype: float64

The dataset is fairly balanced between fake and real news 

In [43]:
# Inspect the date column
data['date'].nunique()

2397

The date column contains 2397 unique dates

## 3. Data Preparation


# Validity

To confirm that the data in each column matches a specified data type and format.


In [44]:
# converting date from object to datetime

data['date'] = pd.to_datetime(data['date'], dayfirst=True, errors='coerce')


In [45]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44898 entries, 0 to 23480
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   title     44898 non-null  object        
 1   text      44898 non-null  object        
 2   subject   44898 non-null  object        
 3   date      44888 non-null  datetime64[ns]
 4   category  44898 non-null  int64         
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 2.1+ MB


In [46]:
data['date'].head()

0   2017-12-31
1   2017-12-29
2   2017-12-31
3   2017-12-30
4   2017-12-29
Name: date, dtype: datetime64[ns]

In [47]:
# extracting months and years

data['year'] = data['date'].dt.year
data['month'] = data['date'].dt.month


In [48]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44898 entries, 0 to 23480
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   title     44898 non-null  object        
 1   text      44898 non-null  object        
 2   subject   44898 non-null  object        
 3   date      44888 non-null  datetime64[ns]
 4   category  44898 non-null  int64         
 5   year      44888 non-null  float64       
 6   month     44888 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(3)
memory usage: 2.7+ MB


In [49]:
# converting the subject column to category

data['subject'] = data['subject'].astype('category')

In [50]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44898 entries, 0 to 23480
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   title     44898 non-null  object        
 1   text      44898 non-null  object        
 2   subject   44898 non-null  category      
 3   date      44888 non-null  datetime64[ns]
 4   category  44898 non-null  int64         
 5   year      44888 non-null  float64       
 6   month     44888 non-null  float64       
dtypes: category(1), datetime64[ns](1), float64(2), int64(1), object(2)
memory usage: 2.4+ MB


# Consistency

To inspect the duplicates

In [51]:
duplicates  = data[data.duplicated()]
duplicates

title  \
445    Senate tax bill stalls on deficit-focused 'tri...   
778    Trump warns 'rogue regime' North Korea of grav...   
892    Republicans unveil tax cut bill, but the hard ...   
896    Trump taps Fed centrist Powell to lead U.S. ce...   
974    Two ex-Trump aides charged in Russia probe, th...   
...                                                  ...   
21353  Thailand's ousted PM Yingluck has fled abroad:...   
21408  U.S., North Korea clash at U.N. forum over nuc...   
9942   HILLARY TWEETS MESSAGE In Defense Of DACA…OOPS...   
11446  FORMER DEMOCRAT WARNS Young Americans: “Rioter...   
14925  [VIDEO] #BlackLivesMatter Terrorists Storm Dar...   

                                                    text       subject  \
445    WASHINGTON (Reuters) - The U.S. Senate on Thur...  politicsNews   
778    BEIJING (Reuters) - U.S. President Donald Trum...  politicsNews   
892    WASHINGTON (Reuters) - U.S. House of Represent...  politicsNews   
896    WASHINGTON (Reuters) - President Donald Trump ...  politicsNews   
974    WASHINGTON (Reuters) - Federal investigators p...  politicsNews   
...                                                  ...           ...   
21353  BANGKOK (Reuters) - Ousted Thai prime minister...     worldnews   
21408  GENEVA (Reuters) - North Korea and the United ...     worldnews   
9942   No time to waste   we've got to fight with eve...      politics   
11446   Who is silencing political speech, physically...      politics   
14925  They were probably just looking for a  safe sp...      politics   

            date  category    year  month  
445   2017-11-30         1  2017.0   11.0  
778   2017-11-08         1  2017.0   11.0  
892   2017-11-02         1  2017.0   11.0  
896   2017-11-02         1  2017.0   11.0  
974   2017-10-30         1  2017.0   10.0  
...          ...       ...     ...    ...  
21353 2017-08-25         1  2017.0    8.0  
21408 2017-08-22         1  2017.0    8.0  
9942  2017-09-09         0  2017.0    9.0  
11446 2017-03-10         0  2017.0    3.0  
14925 2015-11-16         0  2015.0   11.0  

[209 rows x 7 columns]

In [52]:

print(f" before: {len(data)}")
data.drop_duplicates(inplace=True)
print(f" after: {len(data)}")



 before: 44898
 after: 44689


# Completeness

From data understanding section we found there to be no missing values. Therefore we can confirm that the the dataset is complete.

# Uniformity

 To check on whether different systems refer to the same value in the same format

In [53]:
data['subject'].value_counts()

politicsNews       11220
worldnews           9991
News                9050
politics            6838
left-news           4459
Government News     1570
US_News              783
Middle-east          778
Name: subject, dtype: int64

In [59]:
data['subject'].replace({'politicsNews': 'politics',
                         'worldnews': 'world_news',
                         'News': 'news',
                         'left-news': 'left_news',
                         'Government News': 'government_news',
                         'US_News': 'us_news',
                         'Middle-east': 'middle_east'}, inplace=True)


In [60]:
data['subject'].value_counts()

politics           18058
world_news          9991
news                9050
left_news           4459
government_news     1570
us_news              783
middle_east          778
Name: subject, dtype: int64

# Exploratory Data Analysis

In [63]:
data['text'].str.contains('WASHINGTON').sum()

6985